Full review

Timple opened the pull request:
PR for commenting on the complete codebase.

Timple committed:
Robustify against negative dt values

Timple committed:
Merge pull request #1 from nobleo/bugfix/HARVEY-812-negative-dt

Robustify against negative dt values

MCFurry committed:
Make init-vel threshold tunable so users can pick a value for their application

MCFurry committed:
Merge pull request #3 from nobleo/fix/init_vel_threshold_tunable

Make init-vel threshold tunable so users can pick a value for their application

Timple committed:
2.16.0

ros-path_tracking_pid  2.16.0  81b78b9
ros-roslint            0.12.0

Timple committed:
Merge pull request #4 from nobleo/release/2.16.0

2.16.0

Timple committed:
Enable industrial CI

Timple committed:
See if noetic fails as well

Timple committed:
Relax the criterium a bit

Timple committed:
Merge pull request #5 from nobleo/industrial-ci

Industrial ci

MCFurry committed:
Fix backwards cancel bug and add test for it

MCFurry committed:
Fix test!?

MCFurry committed:
Merge pull request #7 from nobleo/fix/HARVEY-855-cancel-on-backward-turn

Fix backwards cancel bug and add test for it

MCFurry committed:
Fix acceleration/deceleration switching in reverse driving

MCFurry committed:
Add ugly test for checking forward and reverse acceleration/deceleration values

MCFurry committed:
Slight test improvements

rokusottervanger commented on src/path_tracking_pid_local_planner.cpp:
```diff
@@ -0,0 +1,644 @@
+    ROS_ERROR_THROTTLE(5, "dt=0 detected, skipping loop(s). Possible overloaded cpu or simulating too fast");
+    cmd_vel = geometry_msgs::Twist();
+    cmd_vel.linear.x = pid_controller_.getControllerState().current_x_vel;
+    cmd_vel.angular.z = pid_controller_.getControllerState().current_yaw_vel;
```
Looking at the controller, I feel like this variable name (`current_x_vel`) as part of the controller state is misleading, as it is really the last velocity setpoint resulting from the controller.

Timple replied:
I think you are right. So `x_vel_setpoint` may be better?

rokusottervanger replied:
Raised a general issue regarding unclear variable naming: #87

rokusottervanger commented on src/controller.cpp:
```diff
@@ -0,0 +1,1230 @@
+    // TODO(clopez/mcfurry/nobleo): Give feedback to higher level software here
+    ROS_WARN("Large control error. Current_x_vel %f / odometry %f",
+            controller_state_.current_x_vel, odom_twist.linear.x);
+  }
```
"`reset current_x_vel`"? I don't see this happening...

rokusottervanger commented on src/controller.cpp:
```diff
@@ -0,0 +1,1230 @@
+  controller_state_.current_x_vel = new_x_vel;
+  controller_state_.current_yaw_vel = new_yaw_vel;
+  return output_combined;
+}
```
This single function is nearly 470 lines of code. Split this up.

Timple replied:
Sounds good. We should probably enable sonarcloud on this repository as well. It gives indications like cyclomatic-complexity of the code. And maximum number of seperate statements in functions.

rokusottervanger replied:
This refactor is in progress. Should I raise an issue about CI missing?

rokusottervanger reviewed (changes requested):
See the comments. Additionally, I'm curious as to what the rationale is behind the separation between the `controller.cpp` file and the `path_tracking_pid.cpp` file. Could you elaborate?

Timple commented on src/path_tracking_pid_local_planner.cpp:
```diff
@@ -0,0 +1,644 @@
+  }
+  ROS_INFO("Finished waiting loop, done cancelling");
+  return true;
+}
```
Hah nice, that implementation in mbf is per our request if I remember correctly. Perhaps we never got around to implement that. 
This part of the code originated before that feature, so can be improved!

Timple replied:
Ah, I did remember correctly! https://github.com/magazino/move_base_flex/pull/172 and https://github.com/magazino/move_base_flex/issues/171

rokusottervanger replied:
See #88

Timple commented on src/path_tracking_pid_local_planner.cpp:
```diff
@@ -0,0 +1,644 @@
+  debug_pub_ = nh.advertise<path_tracking_pid::PidDebug>("debug", 1);
+  path_pub_ = nh.advertise<nav_msgs::Path>("visualization_path", 1, true);
+
+  sub_odom_ = gn.subscribe("odom", 1, &TrackingPidLocalPlanner::curOdomCallback, this);
```
Interesting... How did we miss that.
Sounds like a good thing to improve on!

rokusottervanger replied:
See #89 and #90

Timple commented on src/path_tracking_pid_local_planner.cpp:
```diff
@@ -0,0 +1,644 @@
+void TrackingPidLocalPlanner::velMaxExternalCallback(const std_msgs::Float64& msg)
+{
+  pid_controller_.setVelMaxExternal(msg.data);
+}
```
Yes, but this feature is for high-dynamic (thus streaming) maximum velocities.

If you want a guaranteed reception because of the (semi-)static nature of the limits, you can simply use the dynamic reconfigure (which is a service).

cesar-lopez-mar replied:
Is this high-dynamic situation really needed? is dynamic reconfigure not dynamic-enough for the intended use?

Timple replied:
Unlike most robots, in Harvey we can run at full capacity of the drive-train. Since the power-input has a maximum, we have to limit the power-output. Slowing down is the easiest (and desired) way to accomplish this.

Due to the dynamic nature of our working-force, we adapt the output velocity at a 10hz rate.
Services are typically not meant for these kind or rates.

Timple commented:
As I'm not the original (nor the main) author here, I didn't address all comments :slightly_smiling_face:

MCFurry committed:
Change accel/decel check in test

MCFurry committed:
Try fix test?

MCFurry committed:
This seems to fix the tests?

MCFurry committed:
using sign() function here is not necessary...

MCFurry committed:
sign() function not needed here

rokusottervanger commented on src/controller.cpp:
```diff
@@ -0,0 +1,1230 @@
+      break;
+  }
+
+  // When velocity error is too big reset current_x_vel
+  if (fabs(odom_twist.linear.x - controller_state_.current_x_vel) > max_error_x_vel_)
+  {
+    // TODO(clopez/mcfurry/nobleo): Give feedback to higher level software here
+    ROS_WARN("Large control error. Current_x_vel %f / odometry %f",
+            controller_state_.current_x_vel, odom_twist.linear.x);
+  }
```
"`reset current_x_vel`"? I don't see this happening...

cesar-lopez-mar replied:
There was a reset at some point, but then it introduced a undesired behavior in one of the applications, we should remove or rewrite the comment.

rokusottervanger replied:
We can get rid of this condition and warning, because it is doesn't do anything functionally, and a similar check is performed in the `update` method here:
https://github.com/nobleo/path_tracking_pid/blob/61c6cc828bd98dddd8cc14cbabdd62353a89ad3c/src/controller.cpp#L642-L649

rokusottervanger commented on src/controller.cpp:
```diff
@@ -0,0 +1,1230 @@
+  }
+  ROS_DEBUG("t_end_phase_current: %f", t_end_phase_current);
+  ROS_DEBUG("d_end_phase: %f", d_end_phase);
+  ROS_DEBUG("distance_to_goal: %f", distance_to_goal_);
```
I'm guessing right now that it is the distance to the last pose in the plan appended with a pose 1 carrot distance away from the last pose in the original plan.

cesar-lopez-mar replied:
This is the distance to goal along the path, we should indeed use a different name.

cesar-lopez-mar replied:
There is a mismatch of the behavior of this variable in the two control modes (trackin_base_link), which should be solved. Issue will be created.

rokusottervanger replied:
Problem is now that it is the distance along the path to the last pose in the global plan, which may be appended with another pose depending on the control strategy.

As discussed, candidate improvements are:
 - Do not append the global plan, but rather extrapolate the end at runtime (if necessary based on the specific control strategy at the time), this would solve the bug referred to in this conversation.
 - Maybe we could refactor the calculations in the two different control strategies to reduce code duplication.

rokusottervanger replied:
See #83

rokusottervanger reviewed (commented):
~~I'm losing confidence in the viability of this controller for the long term... It takes a complete reimplementation to fix this.~~

Rayman commented on include/path_tracking_pid/controller.hpp:
```diff
@@ -0,0 +1,368 @@
+
+  // Cutoff frequency for the derivative calculation in Hz.
+  // Negative -> Has not been set by the user yet, so use a default.
+  double cutoff_frequency_long_ = -1.0;
```
Unused private field

Rayman commented on include/path_tracking_pid/controller.hpp:
```diff
@@ -0,0 +1,368 @@
+  double feedforward_ang_ = 0.0;
+  double xvel_ = 0.0;
+  double yvel_ = 0.0;
+  double thvel_ = 0.0;
```
Unused private fields `xvel_`, `yvel_` and `thvel_`

Rayman commented on include/path_tracking_pid/path_tracking_pid_local_planner.hpp:
```diff
@@ -0,0 +1,182 @@
+  {
+    std::vector<polygon_t> output_vec;
+    boost::geometry::union_(polygon1, polygon2, output_vec);
+    return output_vec.at(0);  // Only first vector element is filled
```
The union of two polygons can be two polygons if there is no overlap. So this funciton always throws away the second

Timple replied:
Ah, good point. We expect them to overlap though, so we should probably warn on size > 1

Rayman committed:
Cleanup unused dependencies

Rayman committed:
Fix compile warnings

Timple commented on src/path_tracking_pid_local_planner.cpp:
```diff
@@ -0,0 +1,644 @@
+  nh.param<std::string>("base_link_frame", base_link_frame_, "base_link");
+
+  nh.param<bool>("use_tricycle_model", use_tricycle_model_, false);
+  nh.param<std::string>("steered_wheel_frame", steered_wheel_frame_, "steer");
```
Vehicles rarely get a fourth wheel on runtime :slightly_smiling_face: 

And frame-names are assumed to be static in ROS mostly?

Rayman replied:
For me the cfg file is very useful to get a quick look at which parameters are available. It would be nice if all parameters would be described in it. Even if some parameters won't change on runtime

Timple replied:
Can be done, but then the code should cope with changing them.

Timple commented on src/controller.cpp:
```diff
@@ -0,0 +1,1230 @@
+  tf2::fromMsg(global_plan.back().pose, last_transform);
+  global_plan_tf_.push_back(last_transform);
+
+  if (!track_base_link_enabled_)
+  {
+    // Add carrot length to plan using goal pose (we assume the last pose contains correct angle)
+    tf2::Transform carrotTF;
+    carrotTF.setOrigin(tf2::Vector3(l_, 0.0, 0.0));
+    global_plan_tf_.push_back(last_transform * carrotTF);
+  }
```
> But what if this pose is misaligned with the last path segment?

That would violate the conditions this controller operates in. Although I'm not sure we validate that.

> what if the carrot distance is changed during execution of the plan

Ah yes, that's probably an oversight!

cesar-lopez-mar replied:
This is indeed an issue when following the original path with a carrot, in that case we should only extend the final goal if the end velocity is zero, but of course it can change online, we should have a discussion on what is the best to do.

rokusottervanger replied:
Raised issue #83 for this

Timple commented on src/controller.cpp:
```diff
@@ -0,0 +1,1230 @@
+  // For now do not allow repeated points or in-place rotation
+  // To allow that the way the progress is checked and the interpolation is done needs to be changed
+  // Also check if points suddenly go in the opposite direction, this could lead to deadlocks
+  for (int pose_idx = 1; pose_idx < global_plan.size() - 1; ++pose_idx)
+  {
+    auto prev_pose = global_plan[pose_idx - 1].pose;
+    auto pose = global_plan[pose_idx].pose;
+    auto next_pose = global_plan[pose_idx + 1].pose;
+    // Check if the angle of the pose is obtuse, otherwise warn and ignore this pose
+    // We check using Pythagorean theorem: if c*c > (a*a + b*b) it is an obtuse angle and thus we can follow it
+    double a_squared = distSquared(prev_pose, pose);
+    double b_squared = distSquared(pose, next_pose);
+    double c_squared = distSquared(prev_pose, next_pose);
+    if (c_squared > (a_squared + b_squared))
+    {
+      tf2::fromMsg(pose, transform);
+      global_plan_tf_.push_back(transform);
+    }
+    else
+    {
+      ROS_WARN("Pose %d of path is not used since it is not in the expected direction of the path!", pose_idx);
+    }
+  }
```
> You're going to cut corners along the way

Can you clarify this? I don't think this controller is cutting any corners (if the point-density of the curves is high enough).

cesar-lopez-mar replied:
As @Timple indicated, the controller is not intended for sparse waypoints, but instead a defined path with high enough points (depending on the application). To follow sharp corners, the plan should be executed in different sections.

rokusottervanger replied:
Then this should be a reason to reject the plan, not make up a new one and follow that.

cesar-lopez-mar replied:
This check should be moved to the setPlan and reject the plan. Issue will be open.

rokusottervanger replied:
Opened #86 for this

Rayman committed:
Merge pull request #13 from nobleo/fix/compile-warnings

Fix compile warnings & small cleanup

Rayman committed:
Cleanup header files

Rayman committed:
Refactor visualization to a separate class

PaulVerhoeckx committed:
fix: Use target_x_vel to determine collision look ahead direction

PaulVerhoeckx committed:
fix: Prevent configuring a collision check resolution of zero

PaulVerhoeckx committed:
fix: Prevent dividing by zero when collision_look_ahead_distance is zero

PaulVerhoeckx committed:
Merge pull request #18 from nobleo/fix/prevent-devisions-by-zero

Fix/prevent devisions by zero

MCFurry committed:
Merge pull request #11 from nobleo/fix/HARVEY-939-acc-dec-switched-in-reverse

Fix/harvey 939 acc dec switched in reverse

rokusottervanger committed:
fix: check predicted footprints for LETHAL_OBSTACLE instead of INSCRIBED_INFLATED_OBSTACLE

rokusottervanger committed:
fix: make velocity scaledown consider costmap only at predicted base_link

cesar-lopez-mar commented on include/path_tracking_pid/controller.hpp:
```diff
@@ -0,0 +1,368 @@
+{
+
+#define RADIUS_EPS 0.001  // Smallest relevant radius [m]
+#define VELOCITY_EPS 1e-3  // Neglegible velocity
```
If I remember correctly these definitions were introduced to check "practical" zero values, I would say they remain the same to most typical applications this controller is intended to, e.g. mobile robots in indoor environments and agricultural robots. We could add them as parameters but I do not see a great added value for the user.

rokusottervanger replied:
I'll resolve this for now, and a user can raise an issue if this becomes a problem.

cesar-lopez-mar commented on launch/path_tracking_pid_mbf.launch:
```diff
@@ -0,0 +1,10 @@
+        <rosparam file="$(find path_tracking_pid)/test/local_costmap_params.yaml" command="load" />
+    </node>
+
+</launch>
```
Or perhaps add it as a test dependency?

Timple replied:
This should have been tackled by the roslaunch checker.
Should be an `exec_depend`

rokusottervanger replied:
This type of errors will be caught by a proper CI. Reason to set this up?

cesar-lopez-mar commented on src/path_tracking_pid_local_planner.cpp:
```diff
@@ -0,0 +1,644 @@
+  {
+    ROS_ERROR("Significant diff between odom (%f) and controller_state (%f) detected. Aborting!",
+              latest_odom_.twist.twist.linear.x, pid_controller_.getControllerState().current_x_vel);
+    return false;
```
I think this was introduced when an user cancels an ongoing action and moves the robot manually. Then wen restarted, there might be a big difference between controller internal setpoint state and the current velocity.

rokusottervanger replied:
Sounds like a reason to reset the controller state, not to reject the goal. This is something you can always do on reception of a new goal, right? Simply by using the current state of the robot.

cesar-lopez-mar replied:
This check is only done in one of the operating modes. Documentation, variable names and error messages, needs to be updated

cesar-lopez-mar replied:
Another idea is to abort the execution at the controller step function. This makes that the odometry subscription is not needed anymore because move_base_flex provides it when the controller is active.

rokusottervanger replied:
See #94

cesar-lopez-mar commented on src/controller.cpp:
```diff
@@ -0,0 +1,1230 @@
+  // Force target_end_x_vel at the very end of the path
+  // Or when the end velocity is reached.
+  // Warning! If target_end_x_vel == 0 and min_vel = 0 then the robot might not reach end pose
+  if ((distance_to_goal_ == 0.0 && target_end_x_vel >= VELOCITY_EPS)
+      || (controller_state_.end_phase_enabled && new_x_vel >= target_end_x_vel - VELOCITY_EPS
+          && new_x_vel <= target_end_x_vel + VELOCITY_EPS))
+  {
+    controller_state_.end_reached = true;
+    controller_state_.end_phase_enabled = false;
+    *progress = 1.0;
+    *eda = 0.0;
+    enabled_ = false;
+  }
+  else
+  {
+    controller_state_.end_reached = false;
+    // eda (Estimated duration of arrival) estimation
+    if (fabs(target_x_vel) > VELOCITY_EPS)
+    {
+      double t_const = (copysign(distance_to_goal_, target_x_vel) - d_end_phase) / target_x_vel;
+      *eda = fmin(fmax(t_end_phase_current, 0.0) + fmax(t_const, 0.0), LONG_DURATION);
+    }
+    else
+    {
+      *eda = LONG_DURATION;
+    }
+  }
```
There might be a dead lock if we do so. For instance if we follow the original path using a carrot in front of the robot (track_base_link is false), the final angular and translation error will depend a lot on the shape of the end of the path. So we check the end of the path when the carrot reaches the last point on the path, and not the robot's base.

cesar-lopez-mar commented:
I propose we have a meeting to address the comments of @rokusottervanger and @Rayman, we can also elaborate more on the design choices. I will sent an email about it.

Timple commented on src/controller.cpp:
```diff
@@ -0,0 +1,1255 @@
+    else
+    {
+      //  0.5*dpY*( 1 + dpX*dpX/(dpY*dPY) );
+      // turning_radius_vector[idx_path] = 0.5*(1/dpY)*( dpY*dpY + dpX*dpX );
```
This commented code hasn't run for years. So can be removed I guess.

Timple commented on src/controller.cpp:
```diff
@@ -0,0 +1,1255 @@
+{
+  setPlan(current_tf, odom_twist, global_plan);
+  controller_state_.previous_steering_angle = tf2::getYaw(tf_base_to_steered_wheel.rotation);
+  // TODO(clopez) use steering_odom_twist to check if setpoint is being followed
```
Is this still planned?

Timple commented on src/controller.cpp:
```diff
@@ -0,0 +1,1255 @@
+    // Find closes robot position to path and then project carrot on goal
+    current_pos_on_plan_ = current_goal_ = findPositionOnPlan(current_tf, &controller_state_, path_pose_idx);
+    // To track the base link the goal is then transform to the control point goal
+    double theda_rp = tf2::getYaw(current_goal_.getRotation());
```
This value was already calculated above. Does that need to happen again? And only for this `if` scope?

rokusottervanger commented:
> I propose we have a meeting to address the comments of @rokusottervanger and @Rayman, we can also elaborate more on the design choices. I will sent an email about it.

Great idea! Tim has already elaborated on some, and the principle is starting to grow on me. Implementation-wise it could do with some improvements to make it 
 - More widely applicable
 - Follow interface and behavior standards better to improve interchangeability with other controllers
 - Maybe even less complex, so that a fix to handle one edge case doesn't break other edge cases (or even nominal behavior).

Let's discuss soon!

PaulVerhoeckx committed:
fix: Use absolute largest velocity to determine collision look ahead offset direction

rokusottervanger committed:
test: make tracking error assessment configurable

rokusottervanger committed:
test: disregard final tracking error in lethal obstacle test

Rayman committed:
Merge pull request #14 from nobleo/refactor/cleanup-headers

Cleanup header files

lewie-donckers committed:
Fixed signed/unsigned comparisons.

lewie-donckers committed:
Stricter compiler settings.

lewie-donckers committed:
Switched to enum classes.

lewie-donckers committed:
Replaced macros with constants.

lewie-donckers committed:
Moved constants to reduce scope.

lewie-donckers committed:
Moved helper function.

lewie-donckers committed:
Replaced C-style arrays.

lewie-donckers committed:
Remove unsused member variables.

PaulVerhoeckx committed:
Merge pull request #15 from nobleo/fix/collision-look-ahead-length-offset

fix: Use target x velocity to determine collision look ahead direction

lewie-donckers committed:
Mark overriden member functions as such.

lewie-donckers committed:
Distinguish between system and normal headers.

lewie-donckers committed:
Even stricter compiler settings.

lewie-donckers committed:
Replaced bind with lambda.

lewie-donckers committed:
Cleaned up includes.

lewie-donckers committed:
Explicitly disabled copy/move.

lewie-donckers committed:
Replaced member variables with locals.

lewie-donckers committed:
Replaced member variable with local.

lewie-donckers committed:
Replaced member variables with constants.

lewie-donckers committed:
Ignore Python cache files.

lewie-donckers committed:
Switched to std::clamp().

lewie-donckers committed:
Removed dead code.

lewie-donckers committed:
Remove dead code.

lewie-donckers committed:
Fixed signed/unsigned conversion warnings.

lewie-donckers committed:
Ensure all members are initialized.

lewie-donckers committed:
Added FifoArray abstraction.

lewie-donckers committed:
Added filtered error tracker abstraction.

lewie-donckers committed:
Const locals when possible.

Rayman committed:
Merge pull request #23 from nobleo/refactor/visualization

Refactor visualization to a separate class

lewie-donckers committed:
Use std::exchange().

Timple commented on src/path_tracking_pid_local_planner.cpp:
```diff
@@ -0,0 +1,644 @@
+    marker_pub_.publish(mkControlPose);
+    marker_pub_.publish(mkGoalPose);
+    marker_pub_.publish(mkPosOnPlan);
+  }
```
@Rayman just did this :slightly_smiling_face:

rokusottervanger replied:
I noticed! Very nice :+1:

Rayman committed:
fixup! Refactor visualization to a separate class

Rayman committed:
Let the test fail if exceptions occur in vis_cb

lewie-donckers committed:
Removed config duplicate members.

lewie-donckers committed:
Merge pull request #25 from nobleo/cleanup/fix-signed-unsigned-comparisons

Fixed signed/unsigned comparisons.

cesar-lopez-mar commented:
> 

We would like to separate the controller core code from the plugin wrapper. Eventually we could wrap the controller into a nav2 plugin as well

cesar-lopez-mar commented on src/path_tracking_pid_local_planner.cpp:
```diff
@@ -0,0 +1,644 @@
+      // 'Project' plan by removing z-component
+      pose_stamped.pose.position.z = 0.0;
+    }
+  }
```
There seems to be a discussion about this topic on ROS2. We will check this and decide later on how to proceed.

cesar-lopez-mar commented on src/path_tracking_pid_local_planner.cpp:
```diff
@@ -0,0 +1,644 @@
+    pid_server_->updateConfig(config);
+    lock.unlock();
+    cancel_requested_ =  false;
+  }
```
Good point, to be implemented

rokusottervanger replied:
See #9

Timple committed:
Clarify init_vel_max_diff check

lewie-donckers committed:
Reworked to add_compile_options().

lewie-donckers committed:
Forgot pragma once.
And used ROS clang-format.

lewie-donckers committed:
Merge pull request #26 from nobleo/cleanup/stricter-compiler-settings

Stricter compiler settings.

lewie-donckers committed:
Merge pull request #27 from nobleo/cleanup/switched-to-enum-classes

Switched to enum classes.

lewie-donckers committed:
Reworked formatting.

Rayman committed:
Merge pull request #24 from nobleo/fix/index-error

Fix `IndexError: list index out of range`

lewie-donckers committed:
Merge pull request #28 from nobleo/cleanup/replaced-macros-with-constants

Replaced macros with constants.

MCFurry committed:
Quaternion of carrotTF was not initialized, set to proper value

Rayman commented:
Some compiler warnings are only generated when optimizations are turned on I think. That's because additional code-flow analysis is then needed.

lewie-donckers committed:
Merge pull request #29 from nobleo/cleanup/reduced-scope-of-constants

Moved constants to reduce scope.

lewie-donckers committed:
Merge pull request #30 from nobleo/cleanup/reduced-scope-of-helper-function

Moved helper function.

MCFurry committed:
Fully initialize at once

lewie-donckers committed:
Merge pull request #31 from nobleo/cleanup/replaced-c-style-arrays

Replaced C-style arrays.

lewie-donckers committed:
Merge pull request #32 from nobleo/cleanup/removed-unused-member-variables

Remove unsused member variables.

lewie-donckers committed:
Merge pull request #34 from nobleo/cleanup/mark-overridden-functions

Mark overriden member functions as such.

lewie-donckers committed:
Merge pull request #36 from nobleo/cleanup/distinguish-between-system-and-normal-headers

Distinguish between system and normal headers.

MCFurry committed:
Merge pull request #47 from nobleo/fix/tf2_tf_not_properly_initialized

Quaternion of carrotTF was not initialized, set to proper value

lewie-donckers committed:
Merge pull request #37 from nobleo/cleanup/even-stricter-compiler-settings

Even stricter compiler settings.

lewie-donckers committed:
Merge pull request #39 from nobleo/cleanup/replaced-bind-with-lambda

Replaced bind with lambda.

lewie-donckers committed:
Reworked by moving gitignore.

lewie-donckers committed:
Merge pull request #42 from nobleo/cleanup/replaced-member-variables-with-locals

Replaced member variables with locals.

lewie-donckers committed:
Merge pull request #43 from nobleo/cleanup/replaced-another-member-variable-with-local

Replaced member variable with local.

lewie-donckers committed:
Merge pull request #45 from nobleo/cleanup/ignore-python-cache-files

Ignore Python cache files.

lewie-donckers committed:
Merge pull request #44 from nobleo/cleanup/replaced-member-variables-with-constants

Replaced member variables with constants.

lewie-donckers committed:
Merge pull request #46 from nobleo/feature/use-std-clamp

Switched to std::clamp().

lewie-donckers committed:
Reworked to keep old behavior.

lewie-donckers committed:
Added clang-tidy and fixed all warnings.

lewie-donckers committed:
Reworked by switching to Boost.

lewie-donckers committed:
Replaced helper function sign().

lewie-donckers committed:
Added to_transform() helper.

lewie-donckers committed:
Moved distSquared() helpers.

lewie-donckers committed:
Added is_pose_angle_obtuse() helper.

lewie-donckers committed:
Merge pull request #41 from nobleo/cleanup/explicitly-disable-copy-move

Explicitly disabled copy/move.

lewie-donckers committed:
Merge pull request #50 from nobleo/cleanup/removed-dead-code

Removed dead code.

lewie-donckers committed:
Merge pull request #51 from nobleo/cleanup/removed-more-dead-code

Remove dead code.

lewie-donckers committed:
Merge pull request #52 from nobleo/cleanup/fix-clang-tidy-signed-unsigned-conversion-warnings

Fixed signed/unsigned conversion warnings.

lewie-donckers committed:
Reworked to different defaults.

lewie-donckers committed:
Reworked to use const& for big types.

lewie-donckers committed:
Used ROS2 clang-format on all files.

lewie-donckers committed:
Merge pull request #53 from nobleo/cleanup/initialize-all-members

Ensure all members are initialized.

Rayman committed:
Fix possible unsigned underflow

When the global plan is empty, this comparison could underflow.

Fix #48

lewie-donckers committed:
Reworked to exclude uninitialized variables.

lewie-donckers committed:
Reworked by renaming to SecondOrderLowpass.

lewie-donckers committed:
Merge pull request #55 from nobleo/cleanup/clang-tidy

Added clang-tidy and fixed all warnings.

lewie-donckers committed:
Merge pull request #56 from nobleo/cleanup/use-std-exchange

Use std::exchange().

Rayman committed:
Merge pull request #68 from nobleo/fix/possible-unsigned-underflow

Fix possible unsigned underflow

lewie-donckers committed:
Merge pull request #57 from nobleo/cleanup/add-fifo-array-abstraction

Added FifoArray abstraction.

lewie-donckers committed:
Merge pull request #58 from nobleo/cleanup/add-filtered-error-tracker-abstraction

Added filtered error tracker abstraction.

lewie-donckers committed:
Merge pull request #60 from nobleo/cleanup/const-local-variables

Const locals when possible.

lewie-donckers committed:
Merge pull request #61 from nobleo/cleanup/replaced-helper-function

Replaced helper function sign().

lewie-donckers committed:
Merge pull request #62 from nobleo/cleanup/added-to-transform-helper

Added to_transform() helper.

lewie-donckers committed:
Reworked by switching to distance2().

lewie-donckers committed:
Merge pull request #63 from nobleo/cleanup/moved-dist-squared-helpers

Moved distSquared() helpers.

lewie-donckers committed:
Merge pull request #64 from nobleo/cleanup/added-is-obtuse-helper

Added is_pose_angle_obtuse() helper.

lewie-donckers committed:
Merge pull request #66 from nobleo/cleanup/removed-config-duplicate-members

Removed config duplicate members.

lewie-donckers committed:
Merge pull request #67 from nobleo/cleanup/clang-format

Used ROS2 clang-format on all files.

lewie-donckers committed:
Merge pull request #40 from nobleo/cleanup/cleanup-includes

Cleaned up includes.

lewie-donckers committed:
Extracted filter_plan() function.

lewie-donckers committed:
Added unittests for FifoArray.

lewie-donckers committed:
Reworked by moving test source files.

lewie-donckers committed:
Refactored mandatory ptr parameter to reference.

Rayman committed:
Enable VERBOSE_TESTS

lewie-donckers committed:
Merge pull request #73 from nobleo/cleanup/extract-function-filter-plan

Extracted filter_plan() function.

lewie-donckers committed:
Merge pull request #74 from nobleo/cleanup/unittest_fifo_array

Added unittests for FifoArray.

Rayman committed:
Enable roslint

Rayman committed:
Merge pull request #77 from nobleo/feature/verbose-tests

Enable VERBOSE_TESTS

Rayman committed:
Disable whitespace/braces check

This check was in conflict with the ROS2 clang-format style so I've
disabled it.

Rayman committed:
Fix roslint

Rayman committed:
Merge pull request #78 from nobleo/fix/enable-roslint

Re-enable roslint

cesar-lopez-mar commented on src/controller.cpp:
```diff
@@ -0,0 +1,1230 @@
+  tf2::Transform robot_pose;
+  tf2::convert(current_tf, robot_pose);
+  tf2::Transform base_to_goal = robot_pose.inverseTimes(current_goal_);
+  double angle_to_goal = atan2(base_to_goal.getOrigin().x(), -base_to_goal.getOrigin().y());
```
There is indeed better ways to implement this. Issue will be created.

rokusottervanger replied:
See #82

lewie-donckers committed:
Refactored output parameters of update.

lewie-donckers committed:
Refactored output parameters of distToSegmentSquared().

lewie-donckers committed:
Refactored distToSegmentSquared().

rokusottervanger commented on cfg/Pid.cfg:
```diff
@@ -0,0 +1,65 @@
+
+gen = ParameterGenerator()
+
+gen.add("l", double_t, 0, "Distance between axle and control point", 0.5, -10, 10)
```
See #87

lewie-donckers committed:
Refactored findPositionOnPlan().

lewie-donckers committed:
Split loop into 3.

lewie-donckers committed:
Replaced some loops with algorithms.

lewie-donckers committed:
Reworked by adding documentation.

lewie-donckers committed:
Extracted deltas_of_plan().

lewie-donckers committed:
Reworked by switching to std::clamp().

lewie-donckers committed:
Keep RosLint happy.

lewie-donckers committed:
Extracted distances_to_goal().

lewie-donckers committed:
Extracted inverse_turning_radiuses().

lewie-donckers committed:
Merge pull request #84 from nobleo/cleanup/refactored-output-parameter-to-return-value-2

Refactored output parameters of update.

lewie-donckers committed:
Merge pull request #85 from nobleo/cleanup/refactored-output-parameter-to-return-value-3

Refactored output parameters of distToSegmentSquared().

lewie-donckers committed:
Merge pull request #91 from nobleo/cleanup/refactored-distToSegmentSquared

Refactored distToSegmentSquared().

lewie-donckers committed:
Merge pull request #92 from nobleo/cleanup/refactored-findPositionOnPlan

Refactored findPositionOnPlan().

lewie-donckers committed:
Keep RosLint happy.

lewie-donckers committed:
Merge pull request #76 from nobleo/cleanup/refactored-mandatory-ptr-parameter-to-reference

Refactored mandatory ptr parameter to reference.

lewie-donckers committed:
Refactored setTricycleModel().

lewie-donckers committed:
Refactored setPlan().

lewie-donckers committed:
Refactored findPositionOnPlan().

lewie-donckers committed:
Refactored last Controller functions().

lewie-donckers committed:
Removed unnecessary conversions.

lewie-donckers committed:
Removed global_plan_ member.

Rayman committed:
Fix lowpass & refactor filter structure

- Fixed second order lowpass
- Made the crossover frequency and damping tuneble
- Use the filtered signal for PID calculations
- Add a derivative filter

lewie-donckers committed:
Merge pull request #95 from nobleo/cleanup/refactored-setPlan

Refactored setPlan().

lewie-donckers committed:
Merge pull request #96 from nobleo/cleanup/extracted-setPlan-calculation-1

Extracted deltas_of_plan().

lewie-donckers committed:
Merge pull request #98 from nobleo/cleanup/extracted-setPlan-calculation-2

Extracted distances_to_goal().

lewie-donckers committed:
Merge pull request #99 from nobleo/cleanup/extracted-setPlan-calculation-3

Extracted inverse_turning_radiuses().

lewie-donckers committed:
Moved visualization.

lewie-donckers committed:
Reworked by renaming local variable.

lewie-donckers committed:
Simplified the calculation of goal direction.

lewie-donckers committed:
Keep clang-tidy happy.

lewie-donckers committed:
Disable nav_core::BaseLocalPlanner support.

lewie-donckers committed:
Controller no longer changes plan.

lewie-donckers committed:
Removed unnecessary nav_core references.

lewie-donckers committed:
Merge pull request #109 from nobleo/cleanup/calculations-clang-tidy-fix

Keep clang-tidy happy.

Timple committed:
2.18.0

ros-dynamic_reconfigure  1.7.2   9cd4f8d
ros-path_tracking_pid    2.18.0  5f3c36e
ros-roslint              0.12.0

lewie-donckers committed:
Replaced data member with local.

lewie-donckers committed:
Removed NOLINT.

lewie-donckers committed:
Refactored output parameter.

rokusottervanger committed:
Merge branch 'main' into fix/controller-is-too-conservative-in-its-collision-checking

lewie-donckers committed:
Merge pull request #100 from nobleo/fix/97-refactored-setTricycleModel

Refactored setTricycleModel().

lewie-donckers committed:
Merge pull request #101 from nobleo/fix/97-refactored-setPlan

Refactored setPlan().

lewie-donckers committed:
Merge pull request #102 from nobleo/fix/97-refactored-findPositionOnPlan

Refactored findPositionOnPlan().

lewie-donckers committed:
Merge pull request #103 from nobleo/fix/97-refactored-update

Refactored last Controller functions().

lewie-donckers committed:
Merge pull request #104 from nobleo/fix/97-removed-unnecessary-conversions

Removed unnecessary conversions.

lewie-donckers committed:
Merge pull request #105 from nobleo/fix/97-removed-global-plan-member

Removed global_plan_ member.

lewie-donckers committed:
Reworked by changing logging to error.

Rayman committed:
Refactor collision visualization functions

Rayman committed:
Remove unnecessary marker ids

The markers are already unique because of the ns.

Rayman committed:
Refactor std::transform calls into a separate function

lewie-donckers committed:
Merge pull request #106 from nobleo/fix/38-move-visualization

Moved visualization.

lewie-donckers committed:
Reworked to separate function.

lewie-donckers committed:
Reworked by adding comment.

lewie-donckers committed:
Switched to old GTest syntax for melodic.

lewie-donckers committed:
Fix for melodic that is compatible with noetic.

Rayman committed:
Enable feedforward_ang during testing for lower error

Rayman committed:
Refactor to use color names

Rayman committed:
Add integral filter abstraction

Timple committed:
2.19.0

ros-dynamic_reconfigure  1.7.2
ros-path_tracking_pid    2.19.0  ff653a6
ros-roslint              0.12.0

Rayman committed:
Merge pull request #118 from nobleo/fix/38-move-visualization-refactored

Refactor collision visualization functions

Rayman committed:
Add reset() and configure() tests

lewie-donckers committed:
Merge pull request #108 from nobleo/fix/82-simplify-goal-direction-calculation

Simplified the calculation of goal direction.

lewie-donckers committed:
Merge pull request #110 from nobleo/fix/90-disable-base-local-planner-support

Disable nav_core::BaseLocalPlanner support.

lewie-donckers committed:
Merge pull request #111 from nobleo/fix/86-controller-no-longer-changes-plan

Controller no longer changes plan.

lewie-donckers committed:
Merge pull request #113 from nobleo/cleanup/planner-replaced-data-member-with-local

Replaced data member with local.

lewie-donckers committed:
Merge pull request #114 from nobleo/cleanup/removed-nolint

Removed NOLINT.

lewie-donckers committed:
Merge pull request #115 from nobleo/cleanup/planner-refactor-output-parameter

Refactored output parameter.

rokusottervanger committed:
Merge branch 'main' into fix/controller-is-too-conservative-in-its-collision-checking

Rayman committed:
Increase timeout to reduce test flakeyness

lewie-donckers committed:
Moved distSquared and added tests.

lewie-donckers committed:
Made distToSegmentSquared static.

lewie-donckers committed:
Refactor name and output of distToSegmentSquared.

rokusottervanger committed:
Merge pull request #20 from nobleo/fix/controller-is-too-conservative-in-its-collision-checking

fix: check predicted footprints for LETHAL_OBSTACLE instead of INSCRIBED_INFLATED_OBSTACLE

Timple committed:
Rokus comments

Co-authored-by: Rokus Ottervanger <rokusottervanger@gmail.com>

lewie-donckers committed:
Move, refactor and test closestPointOnSegment().

lewie-donckers committed:
Reworked by renaming data member.

lewie-donckers committed:
Reworked by renaming function.

lewie-donckers committed:
Merge pull request #130 from nobleo/refactor/122-move-distSquared

Moved distSquared and added tests.

lewie-donckers committed:
Merge pull request #131 from nobleo/refactor/122-make-distToSegmentSquared-static

Made distToSegmentSquared static.

lewie-donckers committed:
Merge pull request #132 from nobleo/refactor/122-rename-distToSegmentSquared

Refactor name and output of distToSegmentSquared.

lewie-donckers committed:
Merge pull request #133 from nobleo/refactor/122-move-test-closestPointOnSegment

Move, refactor and test closestPointOnSegment().

Rayman committed:
Merge pull request #119 from nobleo/fix/lowpass

Fix lowpass & refactor filter structure

lewie-donckers committed:
Cleanup Controller::update().

lewie-donckers committed:
Replaced path pose parameter with return value.

lewie-donckers committed:
Made Controller::findPositionOnPlan() const.

lewie-donckers committed:
Replaced distance_to_goal_ data member with local.

lewie-donckers committed:
Moved last visited pose index to find result.

lewie-donckers committed:
Replaced controller_state parameter of findPositionOnPlan().

lewie-donckers committed:
Replaced Controller::getControllerState().

lewie-donckers committed:
Renamed position to pose for findPositionOnPlan.

lewie-donckers committed:
Merge pull request #135 from nobleo/fix/123-cleanup-update

Cleanup Controller::update().

lewie-donckers committed:
Merge pull request #136 from nobleo/fix/123-replace-path-pose-parameter

Replaced path pose parameter with return value.

lewie-donckers committed:
Merge pull request #137 from nobleo/fix/123-made-findPositionOnPlan-const

Made Controller::findPositionOnPlan() const.

lewie-donckers committed:
Merge pull request #138 from nobleo/fix/123-replaced-distance-to-goal-data-member

Replaced distance_to_goal_ data member with local.

lewie-donckers committed:
Merge pull request #139 from nobleo/fix/123-moved-last-visited-pose-index

Moved last visited pose index to find result.

lewie-donckers committed:
Merge pull request #140 from nobleo/fix/123-replaced-controller-state-parameter

Replaced controller_state parameter of findPositionOnPlan().

lewie-donckers committed:
Merge pull request #144 from nobleo/fix/123-renamed-position-to-pose

Renamed position to pose for findPositionOnPlan.

lewie-donckers committed:
Merge pull request #141 from nobleo/fix/124-replace-getControllerState

Replaced Controller::getControllerState().

Rayman committed:
Move getControlPointPose to a separate function & test

Rayman committed:
Simplify closestPoseOnSegment math

PaulVerhoeckx committed:
fix: Use absolute target velocity to calculate end phase distance

PaulVerhoeckx committed:
fix: Enforce positive end phase time and simplify end phase distance

PaulVerhoeckx committed:
Merge pull request #150 from nobleo/fix/end-phase-distance-for-driving-backwards

fix: Use absolute stopping time to calculate end phase distance

Rayman committed:
Merge pull request #148 from nobleo/refactor/closest-pose-on-segment

Move & simplify getControlPointPose

Timple committed:
Refactor projectedCollisionCost to input-output function

Timple committed:
Make projectedCollisionCost static

Timple committed:
Factored out creation of projected footprint

Rayman committed:
Relicense the code under Apache-2.0

The old gpl license might make us vulnerable to oudside contributers, as
discussed in the tech share meeting. Let's relicense under Apache-2.0 to
also address some potential patent related issues.

Relicensing requires the consent of all copyright holders. I’ve tracked
down all contributions, and they were all done by people working at
Nobleo. This means Nobleo is the sole copyright holder.

Rayman committed:
Merge pull request #157 from nobleo/relicense-apache-v2

Relicense the code under Apache-2.0
